receivers/mqtt/mqtt.go

```diff
@@ -0,0 +1,135 @@
+package mqtt
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	mqttLib "github.com/eclipse/paho.mqtt.golang"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/types"
+
+	"github.com/grafana/alerting/logging"
+	"github.com/grafana/alerting/receivers"
+	"github.com/grafana/alerting/templates"
+)
+
+type Client interface {
+	Connect() mqttLib.Token
+	Publish(topic string, qos byte, retained bool, payload interface{}) mqttLib.Token
+	Disconnect(quiesce uint)
+}
+
+type Notifier struct {
+	*receivers.Base
+	log      logging.Logger
+	tmpl     *templates.Template
+	settings Config
+	client   Client
+}
+
+func defaultClientFactory(opts *mqttLib.ClientOptions) Client {
+	return mqttLib.NewClient(opts)
+}
+
+func New(cfg Config, meta receivers.Metadata, template *templates.Template, logger logging.Logger, clientFactory func(opts *mqttLib.ClientOptions) Client) *Notifier {
+	if clientFactory == nil {
+		clientFactory = defaultClientFactory
+	}
+
+	opts := mqttLib.NewClientOptions().
+		AddBroker(cfg.BrokerURL).
+		SetClientID(cfg.ClientID).
+		SetUsername(cfg.Username).
+		SetPassword(cfg.Password)
+
+	if cfg.InsecureSkipVerify {
+		tlsCfg := tls.Config{
+			InsecureSkipVerify: true,
+		}
+		opts.SetTLSConfig(&tlsCfg)
+	}
+
+	return &Notifier{
+		Base:     receivers.NewBase(meta),
+		log:      logger,
+		tmpl:     template,
+		settings: cfg,
+		client:   clientFactory(opts),
+	}
+}
+
+// mqttMessage defines the JSON object send to an MQTT broker.
+type mqttMessage struct {
+	*templates.ExtendedData
+
+	// The protocol version.
+	Version  string `json:"version"`
+	GroupKey string `json:"groupKey"`
+	Message  string `json:"message"`
+}
+
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	n.log.Debug("Sending an MQTT message")
+
+	msg, err := n.buildMessage(ctx, as...)
+	if err != nil {
+		return false, err
+	}
+
+	if token := n.client.Connect(); token.Wait() && token.Error() != nil {
+		n.log.Error("Failed to connect to MQTT broker", "error", token.Error())
+		return false, fmt.Errorf("Failed to connect to MQTT broker: %w", token.Error())
+	}
+
+	if token := n.client.Publish(n.settings.Topic, 0, false, string(msg)); token.Wait() && token.Error() != nil {
+		n.log.Error("Failed to publish MQTT message", "error", token.Error())
+		return false, fmt.Errorf("Failed to publish MQTT message: %w", token.Error())
+	}
+
+	n.client.Disconnect(250)
```

Comment:
1. let's add this as  `defer` right after connection was successful. You do not disconnect in the case of publish error
2. where does 250 come from?

Reply:
1. Updated: https://github.com/grafana/alerting/pull/227/commits/95ac686a86640ac7fd116f06bd230b69d193e7df
2. 250 milliseconds is a completely arbitrary number that I took from the examples and the mqtt-datasource plugin. According to the docs, the client waits for the specified number of milliseconds for the existing work to be completed. In our case, we do not subscribe to any topics, and as far as I understand, the only work it will be waiting for is [the disconnect packet](https://github.com/eclipse/paho.mqtt.golang/blob/714f7c0231294ec4144f0e0e5fc5b43a6d430d2f/client.go#L457), so I'm not sure that we actually need it and can't just put 0 here.